funcr/types/types.go

```diff
@@ -0,0 +1,22 @@
+/*
+Copyright 2022 The logr Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package types provides type definitions for funcr that may also be useful
+// without the funcr implementation.
+package types
+
+// PseudoStruct is a list of key-value pairs that gets logged as a struct.
+type PseudoStruct []interface{}
```

Comment:
why `funcr/types` rather than putting this in `logr` or `logr/x` or similar?

Reply:
Because then we would need to discuss whether this is the right name and type for this kind of thing and how it relates to the upcoming slog changes - I was trying that in https://github.com/go-logr/logr/pull/164.

Moving the type inside funcr sidesteps all of that while still making it easier to support such a feature in Kubernetes (smaller vendor changes). It's a purely pragmatic choice...

Reply:
The problem I have with a more structured form is that all of the Logger interfaces take `[]interface{}`.  `Pseudostruct` was added so that callers could choose to render things like the saved values as a pseudo-struct.  If we normalize on anything but `[]interface{}`, then we are forcing that specific use-case to allocate and/or copy data around.

For example, funcr saves the `WithValues()` values in a `[]interface{}` and offers a hook before rendering them.  That gets called on EVERY SINGLE log line.  If a new `logr.PseudoStruct` is a `[]KeyValuePair`, then that hook needs to make a new slice and populate it, render that, then discard the results (or maybe cache it until it changes, I guess, but that is scary).  That's a lot of work to do on every log line.

So instead we save it as that struct.  That's an API break, so it needs a new hook.  Then any situation which DOESN'T use that hook for this specific purpose has to unpack the slice into a new `[]interface{}` in order to call Info/Error.

It's a loss either way.  funcr could detect which hook you set up and store the data that way natively, I guess.  Yuck.  As I said elsewhere, I am not keen on major changes here, with slog on the horizon.

It feels very weird to have k8s vendor funcr/types and not use funcr itself.

If all you want is the type definition and we are NOT going to make it part of logr proper, why not just copy it into klog?